app/gui/src/presenter/searcher/provider.rs

```diff
@@ -175,32 +277,65 @@ macro_rules! kind_to_icon {
     }
 }
 
-impl list_view::entry::ModelProvider<component_group_view::Entry> for Component {
-    fn entry_count(&self) -> usize {
-        self.group.matched_items.get()
+fn component_to_entry_model(component: &component::Component) -> component_grid::EntryModel {
+    let can_be_entered = component.can_be_entered();
+    let match_info = component.match_info.borrow();
+    let caption = component.label();
+    let highlighted = bytes_of_matched_letters(&*match_info, &caption);
+    let icon = match &component.data {
+        component::Data::FromDatabase { entry, .. } => {
+            let kind = entry.kind;
+            let icon_name = entry.icon_name.as_ref();
+            let icon = icon_name.and_then(|n| n.to_pascal_case().parse().ok());
+            icon.unwrap_or_else(|| for_each_kind_variant!(kind_to_icon(kind)))
+        }
+        component::Data::Virtual { snippet } => snippet.icon,
+    };
+    component_grid::EntryModel {
+        caption: caption.into(),
+        highlighted: Rc::new(highlighted),
+        icon,
+        can_be_entered,
     }
+}
 
-    fn get(&self, id: usize) -> Option<component_group_view::entry::Model> {
-        use model::suggestion_database::entry::for_each_kind_variant;
-        let component = self.group.get_entry(id)?;
-        let is_enterable = component.can_be_entered();
-        let match_info = component.match_info.borrow();
-        let label = component.label();
-        let highlighted = bytes_of_matched_letters(&*match_info, &label);
-        let icon = match component.data {
-            component::Data::FromDatabase { entry, .. } => {
-                let kind = entry.kind;
-                let icon_name = entry.icon_name.as_ref();
-                let icon = icon_name.and_then(|n| n.to_pascal_case().parse().ok());
-                icon.unwrap_or_else(|| for_each_kind_variant!(kind_to_icon(kind)))
-            }
-            component::Data::Virtual { snippet } => snippet.icon,
-        };
-        Some(component_group_view::entry::Model {
-            icon,
-            highlighted_text: list_view::entry::GlyphHighlightedLabelModel { label, highlighted },
-            is_enterable,
-        })
+/// An object providing element models for view's [grid](component_list_panel::grid::View) derived
+/// from [controller's provider](ontroller::searcher::ComponentsProvider).
+///
+/// During construction an FRP network is set up to answer the `model_for_header_needed` and
+/// `model_for_entry_needed` events received from the view. These connections are removed once this
+/// structure is dropped.
+#[derive(Debug)]
+pub struct Component {
```

Comment:
The title "=== Component ===" was placed 200 lines above this declaration. I feel that sections in this file can be structured better.